qiskit/ignis/verification/tomography/experiment/analysis.py

```diff
@@ -0,0 +1,134 @@
+from qiskit.result import Result, Counts
+from qiskit.ignis.verification.tomography.data import marginal_counts, count_keys
+from qiskit.ignis.verification.tomography.basis import TomographyBasis, default_basis
+from qiskit.ignis.verification.tomography.fitters.lstsq_fit import lstsq_fit
+from qiskit.ignis.verification.tomography.fitters.cvx_fit import cvx_fit, _HAS_CVX
+from .utils import _fitter_data
+
+class TomographyAnalysis(Analysis):
+    _HAS_SDP_SOLVER = None
+
+    def __init__(self,
+                 data: Optional[Union[BaseJob, Result, List[any], any]] = None,
+                 metadata: Optional[Union[List[Dict[str, any]], Dict[str, any]]] = None,
+                 meas_basis: Union[TomographyBasis, str] = 'Pauli',
+                 prep_basis: Union[TomographyBasis, str] = 'Pauli',
+                 method: str = 'auto',
+                 name: Optional[str] = None,
+                 exp_id: Optional[str] = None):
+        super().__init__(data, metadata, name, exp_id)
+        self._method = method
+
+        meas_basis = default_basis(meas_basis)
+        if isinstance(meas_basis, TomographyBasis):
+            if meas_basis.measurement is not True:
+                raise QiskitError("Invalid measurement basis")
+        self.meas_matrix_func = meas_basis.measurement_matrix
+
+        prep_basis = default_basis(prep_basis)
+        if isinstance(prep_basis, TomographyBasis):
+            if prep_basis.preparation is not True:
+                raise QiskitError("Invalid preparation basis")
+        self.prep_matrix_func = prep_basis.preparation_matrix
+
+        self._analysis_fn = self.fit
+        self._target_qubits = None
+
+    def set_target_qubits(self, qubits: List[int]):
+        self._target_qubits = qubits
+
+    def _set_target_meta(self, metadata: Dict[str, any]):
+        # restrict meas_qubits, meas_clbits and meas_label to the target qubits
+        target_qubits = self._target_qubits
+        meas_qubits = metadata['meas_qubits']
+        if any([qubit not in meas_qubits for qubit in target_qubits]):
+            raise RuntimeError("Target qubit set {} "
+                               "not contained in measurement "
+                               "qubit set {}".format(target_qubits, meas_qubits))
+        qubit_indices = [meas_qubits.index(qubit) for qubit in target_qubits]
+        metadata['meas_qubits'] = [metadata['meas_qubits'][i] for i in qubit_indices]
+        metadata['meas_clbits'] = [metadata['meas_clbits'][i] for i in qubit_indices]
+        metadata['meas_label'] = [metadata['meas_label'][i] for i in qubit_indices]
+
+    def _format_data(self, data: Result,
+                     metadata: Dict[str, any],
+                     index: int) -> Counts:
+        if self._target_qubits is not None:
+            self._set_target_meta(metadata)
+        meas_qubits = metadata['meas_qubits']
+        counts = data.get_counts(index)
+        counts = marginal_counts(counts, meas_qubits)
```

Comment:
I think this marginalization could be done separately, so it just stores all the raw data first, and then the marginalization is part of the run_analysis. When marginalizing you would want to combine count dicts that have the same marginal bases on the specified qubits for example.

Something like:
```python
def _format_tomography_data(self, qubits=None, clbits=None):
    if clbits is None:
        clbits = qubits
    dataset = {}
    for counts, metadata in zip(self._exp_result, self._exp_metadata):
        if qubits is None:
            prep_label = metadata['prep_label']
            meas_label = metadata['meas_label']
        else:
            prep_label = [metadata['prep_label'][i] for i in qubits]
            meas_label = [metadata['meas_label'][i] for i in qubits]
       mcounts = marginal_counts(counts, clbits)
       key = (prep_label, meas_label)
       if key in dataset:
           mcounts = combine_counts(mcounts, dataset[key])
       dataset[key] = mcounts
   return dataset
```

Then your analysis function could work with this dataset along lines of:

```python
    data, basis_matrix, weights = _fitter_data(dataset, ...)
```